src/sdl2/render.rs

```diff
@@ -125,40 +178,233 @@ impl RendererInfo {
+pub struct Canvas<T> {
+    target: T,
+    context: Rc<RendererContext>,
+}
+
+/// Alias for a `Canvas` that was created out of a `Surface`
+pub type SurfaceCanvas<'s> = Canvas<Surface<'s>>;
+
+/// Methods for the `SurfaceCanvas`.
+impl<'s> Canvas<Surface<'s>> {
+    /// Creates a 2D software rendering context for a surface.
+    ///
+    /// This method should only fail if SDL2 is not built with rendering
+    /// support, or there's an out-of-memory error.
+    pub fn from_surface(surface: surface::Surface<'s>) -> Result<Self, String> {
+        let raw_renderer = unsafe { ll::SDL_CreateSoftwareRenderer(surface.raw()) };
+        if raw_renderer != ptr::null_mut() {
+            let context = Rc::new(unsafe { RendererContext::from_ll(raw_renderer) });
+            Ok(Canvas {
+                   target: surface,
+                   context: context,
+               })
+        } else {
+            Err(get_error())
+        }
+    }
+
+    /// Gets a reference to the associated surface of the Canvas
+    #[inline]
+    pub fn surface(&self) -> &SurfaceRef {
+        &self.target
+    }
+
+    /// Gets a mutable reference to the associated surface of the Canvas
+    #[inline]
+    pub fn surface_mut(&mut self) -> &mut SurfaceRef {
+        &mut self.target
+    }
+
+    /// Gets the associated surface of the Canvas and destroys the Canvas
+    #[inline]
+    pub fn into_surface(self) -> Surface<'s> {
+        self.target
+    }
+
+    /// Sets the render target to the provided texture
+    /// Returns a handle for rendering methods to the target
+    /// Returns `TargetRenderError::NotSupported`
+    /// if the renderer does not support the use of render targets,
+    /// Returns `TargetRenderError::SdlError` if SDL2 returned with an error code.
+    /// The texture must be created with the texture access: `sdl2::render::TextureAccess::Target`.
+    pub fn with_target<'r, 't, 'a>(&'r mut self,
+                                   texture: &'t mut Texture<'a>)
+                                   -> Result<TextureCanvas<'r, 't>, TargetRenderError> {
+        self.internal_with_target(texture)
+    }
+
+    /// Returns a `TextureCreator` that can create Textures to be drawn on this `Canvas`
+    ///
+    /// This `TextureCreator` will share a reference to the renderer and target context.
+    ///
+    /// The target (i.e., `Window`) will not be destroyed and the SDL_Renderer will not be
+    /// destroyed if the `TextureCreator` is still in scope.
+    pub fn texture_creator(&self) -> TextureCreator<SurfaceContext<'s>> {
+        TextureCreator {
+            context: self.context.clone(),
+            _target_context: self.target.context(),
+        }
+    }
+}
+
+pub type WindowCanvas = Canvas<Window>;
+
+/// Methods for the `WindowCanvas`.
+impl Canvas<Window> {
+    /// Gets a reference to the associated window of the Canvas
+    #[inline]
+    pub fn window(&self) -> &WindowRef {
```

Comment:
Why `&WindowRef` and not `&Window` ?
Also, I think `WindowRef` was used for historical reasons, but I don't think it needs to be used anymore, so unless someone can prove me wrong it would be best to "merge" Window and WindowRef as Window.